### PR TITLE
Move navigation and selection logic to WritingFlow

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR moves selection and navigation mode logic (event listeners) to `WritingFlow`.

* This removes a lot of event listeners that are currently added to *each* block.
* The functions are recreated less often since `WritingFlow` is rerendered less often. These functions don't depend on any block data, only on selection state.
* This removes the need for the `IgnoreNestedEvents` component.
* Creates a `RootContainer` component where I moved some common block logic and `useMultiSelection` (which is now only rendering once).
* It simplifies the `BlockListBlock` component. The absence of the event handlers will also make it easier later on to remove the `div` wrappers.
* We may want to refactor `WritingFlow` a little to separate navigation and selection logic.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/block-editor/src/components/block-list/block.js [hunk 1]
```diff
@@ -469,18 +401,15 @@ function BlockListBlock( {
 	);
 
 	return (
-		<IgnoreNestedEvents
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/block.js]: @aduth Might be of interest to you as you originally created this component.

**Comment by youknowriad** [comment]
Did you test the performance impact of this PR. It seems like it has potential for improving the numbers.

**Comment by ellatrix** [comment]
@youknowriad I did a quick check and I think it does improve slightly, but I have to say that, on my computer, the number vary even when I test the master branch a few times in a row. 🤷‍♀️

**Comment by aduth** [comment]
Related (would make redundant): #14707

**Comment by aduth** [comment]
One thing this has me wondering is whether we need to start considering to bake in `WritingFlow` as part of the default rendering of a `BlockEditor`. Right now, it's a separate, optional component. But for a naive rendering of a block editor, the changes here would have the effect that a lot of the interactivity is lost.

I guess we [document this](https://github.com/WordPress/gutenberg/tree/master/packages/block-editor#usage) as more-or-less assumed to be part of a standard rendering hierarchy, so the expectation is already there. Furthermore, we rely on this in the top-level post editor to be able to [include the post title](https://github.com/WordPress/gutenberg/blob/fd31f20b728670b8b5bd578c6c7d27e43136e200/packages/edit-post/src/components/visual-editor/index.js#L32-L39) as part of the consideration of the writing flow.

**Comment by aduth** [comment]
One thing this appears to regress on is that clicking the button block appender in a column block will first select that block before allowing the inserter to be shown. This has been a source of some trouble in the past (e.g. #18329), where the intended behavior is that clicking the appender button should immediately open the inserter.

Steps to reproduce:

1. Navigate to Posts > Add New
2. Insert a Columns block
3. Select any columns arrangement
4. Unselect the Columns block
5. Click one of the appender buttons in any of the columns

Expected: The inserter is opened.
Actual: The block is selected. A second click is required to open the inserter.

Related: #19135, #18379

There are end-to-end in both of those pull requests. I think it might be that the navigation block works correctly (does appear to in my testing, not sure why it's any different). The tests modified in #19135 are primarily to test against regressions to the expected behavior described in #18928. It might be that we lack a test case specifically verifying this columns (button) appender behavior.

**Comment by aduth** [comment]
I generally like the direction this is taking, and am especially fond of the idea of removing the complexities surrounding `IgnoreNestedEvents` in favor of handling block selection events further up the virtual tree. One worry I might have is whether `WritingFlow` is starting to do _too much_, at least in being able to succinctly define its purpose. It's conflicting to me because I do think some of the refactoring could make sense in WritingFlow (e.g. multi-selection, navigation mode toggling). I'm not sure of it all, though, such as the focus handling, at least if we consider WritingFlow as having primarily served the purpose of handling keyboard interactions with respect to their implications on block selection.

Maybe this is a matter of my own unclear expectations around what this component is meant to be responsible for handling. I think it's worth clarifying in any case (ideally, though not necessarily as part of this pull request, as documentation for the component).

Alternatively, it's something where we could consider splitting off some of the specific functionality into separate components. This could be relevant to my previous point about "default" behaviors, where if we don't want to bake-in _all_ of WritingFlow, we could still split out some of the functionality to become default. In my head, I'm thinking of components like [MultiSelectScrollIntoView](https://github.com/WordPress/gutenberg/blob/c24882bc3a81bad12c1c367b2cc9ae982d1a8605/packages/block-editor/src/components/multi-select-scroll-into-view/index.js), which itself is somewhat related to the functionality of `WritingFlow`, but isolated for the purpose of improving clarity around what it's expected to be handling.

**Comment by aduth** [comment]
On the topic of https://github.com/WordPress/gutenberg/pull/19397#issuecomment-570596499 and default behaviors, I wonder if this might have any impact on whether we will be able to implement nested block editors (e.g. #14715, I think also Widgets screen might have some nested editors?). My worry is: How does selection work for those inner editors, if it's meant to be handled by WritingFlow? Does it mean there would need to be multiple WritingFlow, one for each of the BlockEditorProviders?

**Comment by ellatrix** [comment]
I think it might be good to split `WritingFlow` in multiple components, e.g. one for multi-selection, one for navigation, and one for focus handling. Currently `WritingFlow` already does some focus handling for moving it in and out the block.

I also like baking `WritingFlow`, or some parts of it, in the block editor.

I'm sot sure how nested editors would work, and I'm sure we'd have to revise the currently selection logic and WritingFlow as well. Should you be able to select and navigate across editors? If yes, then there should probably be one `WritingFlow` component wrapping everything.

**Comment by ellatrix** [comment]
I'll fix the remaining issue with the column inserter and attempt to split `WritingFlow` in parts.

**Comment by aduth** [comment]
> Should you be able to select and navigate across editors? If yes, then there should probably be one `WritingFlow` component wrapping everything.

I don't think it's necessarily a requirement, especially considering that if each editor is responsible for maintaining its own blocks state, I don't know how a top-level WritingFlow could possibly track the selection for those blocks (since, at that level, the blocks state is not known).

**Comment by ellatrix** [comment]
I've now moved the focus handling to the root `BlockList` so it is backed in. The other stuff like navigation mode and multi selection logic do make sense in `WritingFlow` so I left it there.

File: packages/block-editor/src/components/block-list/index.js [hunk 2]
```diff
@@ -72,18 +71,17 @@ function BlockList( {
 		hasMultiSelection,
 		enableAnimation,
 	} = useSelect( selector, [ rootClientId ] );
-	const ref = useRef();
-	const onSelectionStart = useMultiSelection( { ref, rootClientId } );
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/index.js]: Notice that this is no longer rendered for every block list, but rather only for the root block list.

File: packages/block-editor/src/components/writing-flow/index.js [hunk 3]
```diff
@@ -209,13 +212,53 @@ export default function WritingFlow( { children } ) {
...
+		blockSelectionStart,
+	} = useSelect( selector, [] );
 	const {
 		multiSelect,
 		selectBlock,
 		clearSelectedBlock,
+		setNavigationMode,
 	} = useDispatch( 'core/block-editor' );
 
+	function onMouseDown( event ) {
```
Comment (ellatrix) [inline packages/block-editor/src/components/writing-flow/index.js]: The logic below belongs here in my opinion, as this component handles multi selection as well as navigation mode already. Perhaps we can split this is the future, but a lot of it seems tied together.

File: packages/block-editor/src/components/block-list-appender/index.js [hunk 4]
```diff
@@ -51,26 +54,24 @@ function BlockListAppender( {
...
+			// is clicked. In some browsers (Firefox, Safari), button clicks do
+			// not emit a focus event, which could cause this event to propagate
+			// unexpectedly. The `tabIndex` ensures that the interaction is
+			// captured as a focus, without also adding an extra tab stop.
+			//
+			// See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+			tabIndex={ -1 }
+			// Prevent the block from being selected when the appender is
+			// clicked.
+			onFocus={ stopPropagation }
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list-appender/index.js]: @aduth This fixed the issue you found (https://github.com/WordPress/gutenberg/pull/19397#issuecomment-570602468). Does everything now work correctly?

Comment (aduth) [reply packages/block-editor/src/components/block-list-appender/index.js]: Initial impression: I imagine this could work, yes, considering that `IgnoreNestedEvents` was in many ways a glorified `stopPropagation` anyways. Will test as part of a broader re-review.

File: packages/block-editor/src/components/block-list/use-multi-selection.js [hunk 5]
```diff
@@ -30,35 +35,32 @@ function getDeepestNode( node, type ) {
...
-		return {
-			blockClientIds: getBlockOrder( rootClientId ),
-			isSelectionEnabled: isSelectionEnabled(),
-			isMultiSelecting: isMultiSelecting(),
-			multiSelectedBlockClientIds: getMultiSelectedBlockClientIds(),
-			hasMultiSelection: hasMultiSelection(),
-			getBlockParents,
-		};
-	}
+function selector( select ) {
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/use-multi-selection.js]: Aside: While the separate function declaration still catches me off guard, one thing which is nice here is demonstrating how, if you are able to move it outside the component declaration altogether, there's a safer guarantee that you don't need to provide any dependencies to the `useSelect` call (since there's no longer the component scope from which to inherit dependencies).

File: packages/block-editor/src/components/writing-flow/index.js [hunk 6]
```diff
@@ -209,13 +216,55 @@ export default function WritingFlow( { children } ) {
...
+
+			if ( clientId ) {
+				if ( event.shiftKey ) {
+					if ( blockSelectionStart !== clientId ) {
+						multiSelect( blockSelectionStart, clientId );
+						event.preventDefault();
+					}
+				// Allow user to escape out of a multi-selection to a singular
+				// selection of a block via click. This is handled here since
+				// onFocus excludes blocks involved in a multiselection, as
```
Comment (aduth) [inline packages/block-editor/src/components/writing-flow/index.js]: Nit: This was copied from BlockListBlock, but since "onFocus" doesn't exist in this file, it's a bit confusing to read now.

File: packages/block-editor/src/utils/dom.js [hunk 7]
```diff
@@ -86,3 +86,24 @@ export function isInsideRootBlock( blockElement, element ) {
 export function hasInnerBlocksContext( element ) {
 	return !! element.querySelector( '.block-editor-block-list__layout' );
 }
+
+/**
+ * Finds the block client ID given any DOM node inside the block.
+ *
+ * @param {Node} node DOM node.
+ *
+ * @return {?string} Client ID or undefined if the node is not part of a block.
```
Comment (aduth) [inline packages/block-editor/src/utils/dom.js]: It's an error we've been making for a while now, but `?` indicates a value which can explicitly be _nullable_, not undefined.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#nullable-undefined-and-void-types

This would be more correct as:

```suggestion
 * @return {string|undefined} Client ID or undefined if the node is not part of a block.
```

File: packages/block-editor/src/utils/dom.js [hunk 8]
```diff
@@ -86,3 +86,24 @@ export function isInsideRootBlock( blockElement, element ) {
...
+ */
+export function getBlockClientId( node ) {
+	if ( node.nodeType !== node.ELEMENT_NODE ) {
```
Comment (aduth) [inline packages/block-editor/src/utils/dom.js]: Unless I'm mistaken, this should be:

```suggestion
	if ( node.nodeType !== Node.ELEMENT_NODE ) {
```

https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Constants

Currently, I believe it will always be triggering the condition, since `node.ELEMENT_NODE` would evaluate to `undefined`.

File: packages/block-editor/src/components/block-list/block.js [hunk 9]
```diff
@@ -542,11 +469,10 @@ function BlockListBlock( {
...
 				</Popover>
 			) }
-			<IgnoreNestedEvents
+			<div
 				ref={ blockNodeRef }
-				onDragStart={ preventDrag }
-				onMouseDown={ onMouseDown }
-				onMouseLeave={ onMouseLeave }
+				// Only allow selection to be started from a selected block.
+				onMouseLeave={ isSelected ? onMouseLeave : undefined }
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/block.js]: Is this related to multi-selection? And if it is, would any of it be expected to be handled in WritingFlow instead? Considering you mention it at https://github.com/WordPress/gutenberg/pull/19397/files#r363036970 as being partly responsible for handling multi-selection.

File: packages/block-editor/src/components/block-list-appender/index.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list-appender/index.js]: Thanks! It would be nice the eventually get rid of this as well. I'll look into it sometime separately.

File: packages/block-editor/src/components/block-list-appender/index.js [hunk 10]
```diff
@@ -51,26 +54,24 @@ function BlockListAppender( {
...
+			// force a focus event to occur when an appender `button` element
+			// is clicked. In some browsers (Firefox, Safari), button clicks do
+			// not emit a focus event, which could cause this event to propagate
+			// unexpectedly. The `tabIndex` ensures that the interaction is
+			// captured as a focus, without also adding an extra tab stop.
+			//
+			// See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+			tabIndex={ -1 }
+			// Prevent the block from being selected when the appender is
+			// clicked.
```
Comment (aduth) [inline packages/block-editor/src/components/block-list-appender/index.js]: I was actually surprised to find that in the latest version of the plugin / core version, the behavior of clicking on the inserter in the column block is such that it requires a second click to open. I don't think it's always been like this, since I distinctly remember trying to require only a single click. One potential difference now is that, while it does only require a single click, it doesn't actually select the block. I don't particularly see this as being problematic, and in some ways it's actually an ideal behavior (since the user's intent is to insert, not necessarily to select the block).

**Review by aduth — APPROVED** [review]
I ran through a number of scenarios relating to the affected code. There's a lot of potential impact here, but as best I can tell, it's all in good shape.

While I'm approving, I do think at the very least we should resolve the issue with `Node.ELEMENT_NODE` vs. `node.ELEMENT_NODE`.

It's a little concerning to me how much we rely on the DOM as the source of truth here with the introduction and reliance on `getBlockClientId`, but I also don't think it contradicts how we've approached WritingFlow to date, and it might be worth the compromise for all the benefits/simplification it brings.

File: packages/block-editor/src/components/block-list-appender/index.js (hunk 4 above)
Comment (aduth) [reply packages/block-editor/src/components/block-list-appender/index.js]: > Thanks! It would be nice the eventually get rid of this as well. I'll look into it sometime separately.

Yeah, that's fair. I'm very reluctant about any `stopPropagation`, but I think it's fair to say it's not really "introduced" here, since it was more-or-less the same effect previously.

File: packages/block-editor/src/utils/dom.js (hunk 8 above)
Comment (ellatrix) [reply packages/block-editor/src/utils/dom.js]: I think this is correct. Every node instance has these properties.

**Comment by ellatrix** [comment]
> It's a little concerning to me how much we rely on the DOM as the source of truth here with the introduction and reliance on `getBlockClientId`, but I also don't think it contradicts how we've approached WritingFlow to date, and it might be worth the compromise for all the benefits/simplification it brings.

I agree. I have some future improvements in mind around keeping the block references higher up the tree that could eliminate having to query the DOM. Not sure if and how it will work, but I generally agree that it would be nice to avoid any queries.

File: packages/block-editor/src/components/block-list-appender/index.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list-appender/index.js]: Right, same here. At least it's now explicit that propagation is stopped here. `IgnoreNestedEvents` kind of made it look like it's more ok.

File: packages/block-editor/src/components/block-list/block.js (hunk 9 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block.js]: That's right, this is for multi selection. It would be nice to move this to `WritingFlow` as well. I didn't do it here to avoid a huge PR, as it seems a little more tricky to solve.

File: packages/block-editor/src/utils/dom.js (hunk 8 above)
Comment (aduth) [reply packages/block-editor/src/utils/dom.js]: > I think this is correct. Every node instance has these properties.

Ah, interesting. I did not know this.

File: packages/block-editor/src/components/block-list-appender/index.js (hunk 4 above)
Comment (aduth) [reply packages/block-editor/src/components/block-list-appender/index.js]: > Right, same here. At least it's now explicit that propagation is stopped here. `IgnoreNestedEvents` kind of made it look like it's more ok.

It's a give and take, I think. One nice thing about `IgnoreNestedEvents` is that it wouldn't stop _all_ propagation, only that which it expressly cares about, thus potentially avoiding issues with other (intended) event handling further up the element hierarchy.

File: packages/block-editor/src/components/block-list/root-container.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+export default function RootContainer( { children, className } ) {
+	const ref = useRef();
+	const {
+		selectedBlockClientId,
+		hasMultiSelection,
+	} = useSelect( selector, [] );
+	const { selectBlock } = useDispatch( 'core/block-editor' );
+	const onSelectionStart = useMultiSelection( ref );
```
[the remark this answers is not in the document]
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/root-container.js]: Right, it's memoized with `useCallback`, so it should be fine.

File: packages/block-editor/src/components/block-list/use-multi-selection.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/use-multi-selection.js]: I agree. It's maybe a bit strange to put it outside the component, but why not if you can prevent creating an additional function on every render.

**Comment by aduth** [comment]
>I was actually surprised to find that in the latest version of the plugin / core version, the behavior of clicking on the inserter in the column block is such that it requires a second click to open. I don't think it's always been like this, since I distinctly remember trying to require only a single click. One potential difference now is that, while it does only require a single click, it doesn't actually select the block. I don't particularly see this as being problematic, and in some ways it's actually an ideal behavior (since the user's intent is to insert, not necessarily to select the block).

One thing I've noticed about the new behavior: When inserting a Columns block and immediately proceeding to try to adjust the widths of individual columns, it can now be more difficult to select the columns, where previously it would be enough to click the button inserter, even if it wasn't the intention to actually insert a block. Granted, "Clicking the inserter" for the purpose of selecting a block is itself a bit of an awkward expectation and probably not the desired behavior, but for lack of another option, it was at least more consistent than it is currently.
